ntbs-build/openshift-pipelines/pipelines/ntbs-publish-database.yml

```diff
@@ -0,0 +1,202 @@
+        - name: output
+          workspace: shared-workspace
+    - name: publish-migration
+      params:
+        - name: checkout_directory
+          value: data-migration
+        - name: database
+          value: migration
+        - name: config_map
+          value: $(params.environment)
+        - name: publish_profile_key
+          value: databases.migration-publish-profile
+        - name: reference_dacpacs_path
+          value: ntbs-data-migration/data
+        - name: sql-user-secret
+          value: $(params.environment)-sql-keytab
+      runAfter:
+        - download-migration
+      taskRef:
+        kind: Task
+        name: publish-dacpac
+      workspaces:
+        - name: input
+          workspace: shared-workspace
+    - name: git-clone-specimen-matching
+      params:
+        - name: url
+          value: 'git@github.com:publichealthengland/ntbs-specimen-matching.git'
+        - name: revision
+          value: $(params.branch)
+        - name: subdirectory
+          value: specimen-matching
+      taskRef:
+        kind: Task
+        name: git-clone-with-ssh
+      when:
+        - Input: $(params.deploy-specimen-matching)
+          Operator: in
+          Values:
+            - 'yes'
+      workspaces:
+        - name: output
+          workspace: shared-workspace
+    - name: download-specimen-matching
+      params:
+        - name: database
+          value: specimen-matching
+        - name: container
+          value: specimen-matching
+        - name: commit_hash
+          value: $(tasks.git-clone-specimen-matching.results.commit)
+      runAfter:
+        - git-clone-specimen-matching
+      taskRef:
+        kind: Task
+        name: download-dacpac
+      workspaces:
+        - name: output
+          workspace: shared-workspace
+    - name: publish-specimen-matching
+      params:
+        - name: checkout_directory
+          value: specimen-matching
+        - name: database
+          value: specimen-matching
+        - name: config_map
+          value: $(params.environment)
+        - name: publish_profile_key
+          value: databases.specimen-matching-publish-profile
+        - name: reference_dacpacs_path
+          value: data/DACPAC
+        - name: sql-user-secret
+          value: $(params.environment)-sql-keytab
+      runAfter:
+        - download-specimen-matching
+        - publish-migration
+      taskRef:
+        kind: Task
+        name: publish-dacpac
+      workspaces:
+        - name: input
+          workspace: shared-workspace
+    - name: git-clone-reporting
+      params:
+        - name: url
+          value: 'git@github.com:publichealthengland/ntbs-reporting.git'
+        - name: revision
+          value: $(params.branch)
+        - name: subdirectory
+          value: reporting
+      taskRef:
+        kind: Task
+        name: git-clone-with-ssh
+      when:
+        - Input: $(params.deploy-reporting)
+          Operator: in
+          Values:
+            - 'yes'
```

Comment:
Just so I understand the flow of steps:
1. this "when" will decide whether the git-clone happens
2. the download step has a runAfter: git-clone. Does this mean if the when clause fails, the runAfter won't see the clone step as having run?
Guessing this is the case, just checking!

Reply:
Essentially yes - if this step is excluded by the `when` condition then all dependent steps will also be excluded.

Reply:
As we've just discussed, at the moment if the `migration` deployment is skipped then the downstream deployments will also be skipped. I'll have a look to see if there's a way around this - we do need the publish steps to happen sequentially.

Reply:
It turns out that this is not really possible at the moment - the setting which determines whether `when` clauses affect only tasks, or their dependent tasks as well, is a global one across the whole OpenShift installation (so affecting all PHE projects on this cluster).

Reply:
I think the best we can do is to keep the three deployments independent of one another, and if a release requires releasing in a particular order then we should (a) make sure this is clear in the release instructions, and (b) release by running the pipeline multiple times.